radixdlt-java-common/src/main/java/com/radixdlt/utils/functional/Failure.java

```diff
@@ -23,58 +23,71 @@
 /**
  * Simplest failure descriptor.
  */
-public final class Failure {
-	private final String message;
+public interface Failure {
+	String message();
 
-	private Failure(String message) {
-		this.message = message;
+	int code();
+
+	default <T> Result<T> result() {
+		return Result.fail(this);
+	}
+
+	default <T> Failure with(Object... params) {
+		return failure(code(), message(), params);
 	}
 
 	/**
-	 * Create instance of Failure with given message.
+	 * Create instance of Failure with given code, message format string and parameters.
 	 *
-	 * @param message failure message
+	 * @param code format string
+	 * @param format message format string
+	 * @param values parameters
 	 *
 	 * @return created instance of Failure
+	 *
+	 * @see MessageFormat for supported format string options
 	 */
-	public static Failure failure(final String message) {
-		return new Failure(message);
+	static Failure failure(int code, String format, Object... values) {
+		return failure(code, MessageFormat.format(format, values));
 	}
 
 	/**
-	 * Create instance of Failure with given message and parameters.
+	 * Create instance of Failure with given message and code.
 	 *
-	 * @param format format string
-	 * @param values parameters
+	 * @param message failure message
 	 *
 	 * @return created instance of Failure
-	 *
-	 * @see MessageFormat for supported format string options
 	 */
-	public static Failure failure(final String format, Object... values) {
-		return new Failure(MessageFormat.format(format, values));
-	}
+	static Failure failure(int code, String message) {
+		return new Failure() {
+			@Override
+			public String message() {
+				return message;
+			}
 
-	public String message() {
-		return message;
-	}
+			@Override
+			public int code() {
+				return code;
+			}
 
-	@Override
-	public int hashCode() {
-		return Objects.hash(message);
-	}
+			@Override
+			public int hashCode() {
+				return Objects.hash(message);
+			}
 
-	@Override
-	public boolean equals(final Object obj) {
-		if (obj == this) {
-			return true;
-		}
+			@Override
+			public boolean equals(Object obj) {
+				if (obj == this) {
+					return true;
+				}
 
-		return (obj instanceof Failure) && Objects.equals(((Failure) obj).message(), message);
-	}
+				return (obj instanceof Failure) && Objects.equals(((Failure) obj).message(), message);
```

Comment:
This should check class is the same as well as `code`. Also, weird to have anonymous class have equals, better to just create separate class.

Reply:
Moved into separate class.